### DISCUSSION TITLE
Worker on Linux

### DISCUSSION THREAD

**Discussion by pol2095** [airsdk/Adobe-Runtime-Support#1262] [Q&A] [answered]
Hello,

is-it possible to use a worker on Linux (Ubuntu 20.04) ?

I have this message error :
```
TypeError: Error #1009: Cannot access a property or method of a null object reference.
	at com.worker::Worker/onMainToBack()
	at flash.system::MessageChannel/internalAddEventListener()
	at flash.system::MessageChannel/addEventListener()
	at com.worker::Worker()
[xcb] Unknown sequence number while processing queue
[xcb] Most likely this is a multi-threaded client and XInitThreads has not been called
[xcb] Aborting, sorry about that.
adl: ../../src/xcb_io.c :260 : poll_for_event:  the assertion « !xcb_xlib_threads_sequence_lost » failed.
```
Thanks

**Comment by ajwfrost** [comment]
Which build of the SDK is this? Just looking at the xcb error -> XInitThreads was being called in 33.1.1.633 but we just took that out as it caused other issues that are related to dialogs/menus. And if this is running via ADL then it may be something similar I guess as this error would be popped up in the debugger message dialog...

In terms of the actual question: I thought all this was implemented properly. The null reference is happening within user code (`com.worker.Worker::onMainToBack`) so are you able to confirm what it is that's null? We'll double-check with some of the standard MessageChannel test cases to ensure it's okay on Ubuntu 20..

thanks

**Comment by pol2095** [comment]
Thanks, the null reference was a problem in my code, it's fixed.

but I have this error when I start the worker
```
[xcb] Unknown sequence number while processing reply
[xcb] Most likely this is a multi-threaded client and XInitThreads has not been called
[xcb] Aborting, sorry about that.
MyApp: ../../src/xcb_io.c :641 : _XReply:  the assertion « !xcb_xlib_threads_sequence_lost » failed.
```
but the problem doesn't happen every time, sometimes the worker work correctly

sometimes, the app is black too.

**Comment by pol2095** [comment, marked as answer]
You can reproduce this problem ?

**Reply by ajwfrost** [reply]
Yes, sorry for the delay -> we were a little concerned by the reference to XInitThreads, since putting that call in actually breaks some other stuff.. but it looks like this is actually a different error, where the worker is being set up with a window whereas of course it's not meant to be! I'd hope we have a fix for this one soon..

**Comment by pol2095** [comment]
Thanks, Worker work now on Linux.

I have a strange bug using ASC 2.0, this bug has occured on Windows and Linux sometimes, I use an embed swf that I use as a back worker, if the embed swf is in a swc, sometimes the embed swf is not integrated in the final swf ?
swf final size sometimes 2500 ko, sometimes 2000 ko (embed swf size 500 ko)
This bug never occur using flex sdk or when I use directly my class (without swc).
It's difficult to reproduce

**Reply by ajwfrost** [reply]
What's the command you're using to link in the SWC file? I think it would need to use `include-libraries` so that the contents are always included without you having to actually specify the usage in the main AS file..

**Comment by pol2095** [comment]
I use "mxmlc ... -library-path+=libs/worker.swc ..."

**Reply by ajwfrost** [reply]
So that may be the issue:
> When you use the `library-path` option, the compiler includes only those classes required at compile time in the SWF file, so the entire contents of a library are not necessarily compiled into the SWF file. The `include-libraries` option includes the entire contents of the library, regardless of which classes are required.

If you are using the `library-path` mechanism then it would rely on there being a reference to the class within the main class (or a class referenced from the main class) in order to be considered for inclusion. If you're going to use this as a worker though, then it definitely needs to be included!

Curious that you're not just embedding the worker into the main swf as data though? In fact, I'm very curious how you're actually kicking off the loader if you've only got a class reference or similar .. or are you using the main swf as the worker and then calling into the SWC library at runtime?

The below says that you always need to have a bytearray with the SWF data in it, so having a SWC seems to imply you're just using the main swf (via `this.loaderInfo.bytes`) and then doing something with the SWC library when you hit the `!Worker.current.isPrimordial` condition..? 
https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/system/Worker.html

**Comment by pol2095** [comment]
my worker class in the swc

```
[Embed(source="../../../BackWorker.swf", mimeType="application/octet-stream")]
private static var BackWorkerClass:Class;

private var worker:Worker;

protected var mainToBack:MessageChannel;
protected var backToMain:MessageChannel;

/**
* Start the app in main thread or worker thread mode
*/
public function MainWorker()
{
	//Create worker from the main swf 
	var myWorker:ByteArray = new BackWorkerClass();
	worker = WorkerDomain.current.createWorker(myWorker, true);
	
	//Create message channel TO the worker
	mainToBack = Worker.current.createMessageChannel(worker);
	
	//Create message channel FROM the worker, add a listener.
	backToMain = worker.createMessageChannel(Worker.current);
	backToMain.addEventListener(Event.CHANNEL_MESSAGE, onBackToMain);
	
	//Now that we have our two channels, inject them into the worker as shared properties.
	//This way, the worker can see them on the other side.
	worker.setSharedProperty("backToMain", backToMain);
	worker.setSharedProperty("mainToBack", mainToBack);
	
	//Lastly, start the worker.
	worker.start();
}
```

I don't know if the problem can occur without a worker

**Reply by ajwfrost** [reply]
So as long as your main AS file has got a reference to "MainWorker", then I would expect it should all be linked in properly...
Are you able to send us a copy of the main SWF when it fails to link this properly? and maybe when it does? and we can see if there are other clues in there....

thanks

**Comment by pol2095** [comment]
You can download it [here](http://pol2095.free.fr/worker/bugs/worker.zip)
sometimes 3276 ko, sometimes 2052 ko

**Reply by ajwfrost** [reply]
Okay so fascinatingly, the code import of the SWF file is working fine: the ActionScript bytecode for both of these main swf files has the reference to the embedded Worker swf and this `MyWorkerClass` is then being cast to ByteArray and used in the Worker constructor:

```
     static function com.worker::MyWorker$cinit():*	/* disp_id=0 method_id=11 nameIndex = 0 */
     {
       // local_count=1 max_scope=1 max_stack=2 code_len=14
       // method position=1524 code position=2443
       0    getlocal0     	
       1    pushscope     	
       2    debugfile     	"..Air/bugs/worker/worker_swc/libs/src;com/worker;MyWorker.as"  //stringIndex = 91
       4    getlex        	BackWorker_swf$12e1fb923895098dacbf52b74521fa7d1594463301 //nameIndex = 70
       6    findproperty  	private::MyWorkerClass //nameIndex = 49
       8    debugline     	17
       10   swap          	
       11   setproperty   	private::MyWorkerClass //nameIndex = 49
       13   returnvoid    	
     }
```

But the buggy one does not then have the actual byte array (`DefineBinaryData` tag) included in it (or the AS3 linkage), which is where the problem lies...

We should be able to put together a test case for this and then look into the Falcon source code to see why it may be happening.
(For interest... just found a useful site here: I didn't actually realise that the newer ActionScript compiler was also open-sourced and donated to Apache! https://cwiki.apache.org/confluence/display/FLEX/Falcon%2BOverview - sounds like they had been planning to add in all the necessary MXML support so it could replace the older Flex compiler, which would have made our AIR SDK deployments a lot easier....)

**Comment by pol2095** [comment]
Hello, I have always the same ASC 2.0 problem on `Windows` :
sometimes the main swf size is `14445294 o` and it work, sometimes the main swf size is `13166573 o` and it doesn't work
I use a swc (size is `1286094 o`) that embed a swf (size is `1309339 o`)
the bug occurs using `mxmlc -library-path+=libs/mySwc.swc` and not using `mxmlc -include-libraries+=libs/mySwc.swc`
thanks

**Reply by pol2095** [reply]
UP
can you see the ASC 2.0 intermittent bug using `-library-path` with an embed swf in a swc on Windows 11 ? thanks

swc class :
```
[Embed(source="./BackWorker.swf", mimeType="application/octet-stream")]
private static var MyWorkerClass:Class;

private var worker:Worker;
public function WorkerExample()
{
	var myWorker:ByteArray = new MyWorkerClass();
	if( Worker.isSupported )
	{
		worker = WorkerDomain.current.createWorker(myWorker);
		...
	}
}
```

Main class :
```
private var example:WorkerExample;
public function Main()
{
	example = new WorkerExample();
	...
}
```
